File: 0049-group-anagrams/main.cpp
// #include <algorithm>
// https://stackoverflow.com/questions/9107516/sorting-characters-of-a-c-string
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string,vector<string>> hashmap;
        for (int i = 0; i < strs.size(); i++) {
            string temp = strs[i];
            sort(temp.begin(), temp.end());
            hashmap[temp].push_back(strs[i]);
        }
        vector<vector<string>> result;
        for (auto& it: hashmap) {
            result.push_back(it.second);
        }
        return result;
    }
};

File: 0049-group-anagrams/main.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # hashmap and sort
        # time: O(n * mlogm)
        # space: O(n * m)
        hashmap = {}
        for i in strs:
            sorted_string = ''.join(sorted(i))
            if sorted_string not in hashmap:
                hashmap[sorted_string] = []
            hashmap[sorted_string].append(i)
        res = []
        for k,v in hashmap.items():
            res.append(v)
        return res
